compiler/rustc_hir_analysis/src/check/method/suggest.rs

```diff
@@ -1201,6 +1204,69 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         false
     }
 
+    fn suggest_range_for_iter(
+        &self,
+        tcx: TyCtxt<'tcx>,
+        actual: Ty<'tcx>,
+        source: SelfSource<'tcx>,
+        span: Span,
+        item_name: Ident,
+        ty_str: &str,
+    ) -> bool {
+        if let SelfSource::MethodCall(expr) = source {
+            let mut search_limit = 5;
+            for (_, parent) in tcx.hir().parent_iter(expr.hir_id) {
+                search_limit -= 1;
+                if search_limit == 0 {
+                    break;
+                }
+
+                if let Node::Expr(parent_expr) = parent && is_range_literal(parent_expr) {
+                    let span_included = match parent_expr.kind {
+                            hir::ExprKind::Struct(_, eps, _) =>
+                                eps.len() > 0 && eps.last().map_or(false, |ep| ep.span.contains(span)),
+                            // `..=` desugars into `::std::ops::RangeInclusive::new(...)`.
+                            hir::ExprKind::Call(ref func, ..) => func.span.contains(span),
+                            _ => false,
+                    };
+
+                    if !span_included {
+                        continue;
+                    }
+
+                    let range_def_id = self.tcx.lang_items().range_struct().unwrap();
```

Comment:
this always searches methods for `Range<T>` but the code above seems to also accept `RangeTo` and others (which implement different traits). Can you either change this code to use the correct range type or limit this help to range?

You should be able to get the type of the expr by using `typeck_results.opt_node_type` and then match on it, check that it's an `Adt` and then you can call `bound_type_of` on the `def_id` of the adt type

Reply:
I tried to use `self.typeck_results.borrow().node_type_opt(parent_expr.hir_id)` and the return value is `None`.

I guess because we are at the point of error-handling part of type checking for the right bound of the ranges....

I use updated the code to get the right `lang-item` from `parent_expr.kind`, and add a testcase for it.